### DISCUSSION TITLE
Trying to run FastAPI’s tests in a virtualenv: TypeError while collecting tests?

### DISCUSSION THREAD

**Discussion by musicinmybrain** [fastapi/fastapi#13057] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
git checkout https://github.com/tiangolo/fastapi.git
cd fastapi
python3.13 -m venv _e
. _e/bin/activate
python -m pip install --upgrade pip
pip install -r requirements-tests.txt
pip install -e '.[all]'
bash scripts/test-cov-html.sh
```

### Description

```
[…]
ERROR tests/test_validate_response_recursive/test_validate_response_recursive_pv2.py - TypeError: ('parser', <class 'module'>)
ERROR tests/test_webhooks_security.py - TypeError: ('parser', <class 'module'>)
ERROR tests/test_ws_dependencies.py - TypeError: ('parser', <class 'module'>)
ERROR tests/test_ws_router.py - TypeError: ('parser', <class 'module'>)
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 445 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
============================================================================================= 445 errors in 51.22s =============================================================================================
```

All of the errors during collection have the same `TypeError: ('parser', <class 'module'>)`.

### Operating System

Linux

### Operating System Details

Fedora 41

### FastAPI Version

0.115.6

### Pydantic Version

2.10.3

### Python Version

Python 3.13.0

### Additional Context

Am I doing something wrong here? I didn’t have trouble running FastAPI’s tests in a virtualenv in the past, but something seems to have changed in the past few weeks or so.

Fortunately, I’m not experiencing the same problem running tests in Fedora’s [`python-fastapi` package](https://src.fedoraproject.org/rpms/python-fastapi), which I maintain, but my inability to reproduce issues I do encounter there in a virtualenv means I can’t usefully report them.

(While I had to check the “I promise to help answer people’s discussion questions or review PR’s” button in order to submit this issue, I didn’t mean it. I’m sympathetic to the overwhelming volume of questions here, but I think it’s reasonable and *helpful* to ask occasional questions and report issues as a downstream Linux distribution package maintainer without having to “pay it back” by getting involved in upstream triage.)

**Comment by jgould22** [comment, marked as answer]
All the tests pass for me on python3.12 but in a container with 3.13 I do also get that error. 

My guess is something in anyio[trio]==3.7.0 doesnt like python 3.13.

In 4.5 they mention 
> Added support for the from_uri(), full_match(), parser methods/properties in anyio.Path, newly added in Python 3.13 

So likely because you have 3.13 you getting caught by one of these guys 

https://github.com/agronholm/anyio/pull/740/files#diff-9b0ada949bf00d8698315dbb297a59d14aa291fe7b2405a7bd335d0a7e66d356R361

Looks like if you bump anyio to `4.5` four of the tests fail so the answer is probably for those tests to be debugged and then for anyio to be bumped

The shortest term fix is of course to use 3.12 if that is possible.

**Reply by musicinmybrain** [reply]
Thanks for the input.

> Looks like if you bump anyio to `4.5` four of the tests fail so the answer is probably for those tests to be debugged and then for anyio to be bumped

Hmm. It’s a good suggestion, but I still see the same errors in test collection after doing `pip install anyio==4.5.0`. I wonder why it seems to work for you, but not for me.

I also note that Fedora still has [`python-anyio` 3.7.1](https://src.fedoraproject.org/rpms/python-anyio), and the tests for [`python-fastapi` package](https://src.fedoraproject.org/rpms/python-fastapi) work fine *in the Fedora RPM build environment*. The difference could be due to a different version or downstream patch in some third package other than `python-anyio` or `python-fastapi`, I suppose.

> The shortest term fix is of course to use 3.12 if that is possible.

Indeed, everything works fine if I repeat the original instructions to reproduce this, but use `python3.12` instead of `python3.13`. Unfortunately, that doesn’t help me much, since I’m asking this question in the context of maintaining Fedora’s [`python-fastapi` package](https://src.fedoraproject.org/rpms/python-fastapi), Fedora 41’s system Python is 3.13, and I’m trying to reproduce Python 3.13-specific regressions.

**Reply by jgould22** [reply]
Is the `python-anyio` package installed with trio extra ? 

Also I was just being quick the tests so I might have misread the number.

**Reply by musicinmybrain** [reply]
> Is the `python-anyio` package installed with trio extra ?

Since you said `python-anyio`, I assume you mean in the Fedora package. Yes, the dependencies are based on `requirements-docs.txt`, so the tests run with `anyio[trio]` installed just as they do in a virtualenv.

**Reply by jgould22** [reply]
Yeah that is what I meant, I am not too familiar with distro packaging.

**Comment by musicinmybrain** [comment]
I see that Python 3.13 is still not in the CI test matrix,

https://github.com/fastapi/fastapi/blob/e90af0c0f91513fd41b9f02c31de0a9a7fa16492/.github/workflows/test.yml#L50-L55

which is kind of a shame since it was released a couple of months ago.

**Comment by musicinmybrain** [comment]
Today, this works as expected:

```
git checkout https://github.com/tiangolo/fastapi.git
cd fastapi
uv venv _e
. _e/bin/activate
uv pip install -r requirements.txt
bash scripts/test-cov-html.sh
```

so I’m going to close this dicussion.
